File: src/main/java/no/nav/fo/veilarbregistrering/oppgave/db/OppgaveRepositoryImpl.java
package no.nav.fo.veilarbregistrering.oppgave.db;

import no.nav.fo.veilarbregistrering.bruker.AktorId;
import no.nav.fo.veilarbregistrering.oppgave.OppgaveImpl;
import no.nav.fo.veilarbregistrering.oppgave.OppgaveRepository;
import no.nav.fo.veilarbregistrering.oppgave.OppgaveType;
import no.nav.sbl.sql.SqlUtils;
import org.springframework.jdbc.core.JdbcTemplate;

import java.sql.Timestamp;
import java.time.LocalDateTime;

public class OppgaveRepositoryImpl implements OppgaveRepository {

    private final JdbcTemplate jdbcTemplate;

    public OppgaveRepositoryImpl(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    @Override
    public long opprettOppgave(
            AktorId aktorId,
            OppgaveType oppgaveType,
            long eksternOppgaveId) {

        long id = nesteFraSekvens();
        SqlUtils.insert(jdbcTemplate, "OPPGAVE")
                .value("ID", id)
                .value("AKTOR_ID", aktorId.asString())
                .value("OPPGAVETYPE", oppgaveType.name())
                .value("EKSTERN_OPPGAVE_ID", eksternOppgaveId)
                .value("OPPRETTET", Timestamp.valueOf(LocalDateTime.now()))
                .execute();

        return id;
    }

    private long nesteFraSekvens() {
        return jdbcTemplate.queryForObject("select OPPGAVE_SEQ.nextval from dual", Long.class).longValue();
    }

    @Override
    public OppgaveImpl hentOppgaveFor(AktorId aktorId) {
        String sql = "SELECT * FROM OPPGAVE WHERE AKTOR_ID = ?";
        return jdbcTemplate.queryForObject(sql, new Object[]{aktorId.asString()}, (rs, row) -> {
            return new OppgaveImpl(
                    rs.getLong("ID"),
                    AktorId.valueOf(rs.getString("AKTOR_ID")),
                    OppgaveType.valueOf(rs.getString("OPPGAVETYPE")),
                    rs.getLong("EKSTERN_OPPGAVE_ID"),
                    rs.getTimestamp("OPPRETTET").toLocalDateTime());
        });
    }
}

File: src/main/kotlin/no/nav/fo/veilarbregistrering/oppgave/db/OppgaveRepositoryImpl.kt
package no.nav.fo.veilarbregistrering.oppgave.db

import no.nav.fo.veilarbregistrering.oppgave.OppgaveRepository
import no.nav.fo.veilarbregistrering.bruker.AktorId
import no.nav.fo.veilarbregistrering.oppgave.OppgaveType
import no.nav.fo.veilarbregistrering.oppgave.OppgaveImpl
import org.springframework.jdbc.core.RowMapper
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import java.sql.Timestamp
import java.time.LocalDateTime

class OppgaveRepositoryImpl(private val db: NamedParameterJdbcTemplate) : OppgaveRepository {
    override fun opprettOppgave(
        aktorId: AktorId,
        oppgaveType: OppgaveType,
        oppgaveId: Long
    ): Long {
        val id = nesteFraSekvens()
        val params  = mapOf(
            "id" to id,
            "aktor_id" to aktorId.aktorId,
            "oppgavetype" to oppgaveType.name,
            "ekstern_oppgave_id" to oppgaveId,
            "opprettet" to Timestamp.valueOf(LocalDateTime.now())
        )

        val sql = "INSERT INTO $OPPGAVE" +
                " ($ID, $AKTOR_ID, $OPPGAVETYPE, $EKSTERN_OPPGAVE_ID, $OPPRETTET)" +
                " VALUES (:id, :aktor_id, :oppgavetype, :ekstern_oppgave_id, :opprettet)"

        db.update(sql, params)
        return id
    }

    private fun nesteFraSekvens(): Long {
        val sql = "SELECT nextVal('OPPGAVE_SEQ')"
        return db.queryForObject(sql, emptyMap<String, Any>(), Long::class.java)!!
    }

    override fun hentOppgaverFor(aktorId: AktorId): List<OppgaveImpl> {
        val sql = "SELECT * FROM $OPPGAVE WHERE $AKTOR_ID = :aktor_id"
        return db.query(sql, mapOf("aktor_id" to aktorId.aktorId), rowMapper)
    }

    companion object {
        private const val ID = "ID"
        private const val AKTOR_ID = "AKTOR_ID"
        private const val OPPGAVETYPE = "OPPGAVETYPE"
        private const val EKSTERN_OPPGAVE_ID = "EKSTERN_OPPGAVE_ID"
        private const val OPPRETTET = "OPPRETTET"
        private const val OPPGAVE = "OPPGAVE"

        private val rowMapper = RowMapper { rs, _ ->
            OppgaveImpl(
                rs.getLong(ID),
                AktorId(rs.getString(AKTOR_ID)),
                OppgaveType.valueOf(rs.getString(OPPGAVETYPE)),
                rs.getLong(EKSTERN_OPPGAVE_ID),
                rs.getTimestamp(OPPRETTET).toLocalDateTime()
            )
        }
    }
}